### ISSUE TITLE
Gauntlet legends: HLE gfx doesn't boot

### ISSUE THREAD

**Issue by olivieryuyu** [gonetz/GLideN64#111]
Boots in LLE with PJ64 2.1 with Hatcat RSP

**Comment by Papermanzero** [comment]
Can anyone post the configs, to run this game?
Also with Hatcats RSP it's not working in my case.

**Comment by purplemarshmallow** [comment]
works for me with this config: PJ64, default settings + LLE gfx enabled, default plugins + GLideN64

**Comment by fzurita** [comment]
I made a pull request with the hack taken from Glide64 that lets this game boot. I tested and it seems to work fine with HLE.

**Comment by olivieryuyu** [comment]
the hack will certainly break the graphics in game. there is a odd command here i guess which was not decipher yet.

**Comment by fzurita** [comment]
Interesting, I wonder how Glide64 is managing.

**Comment by olivieryuyu** [comment]
glide64 has error, mostly a kind of flickering in game which cannot be avoided. All HLE plugin currently has this issue.

**Comment by olivieryuyu** [comment]
it is believed that it is due to a yielding issue.

The manual explains some matters about this here: 

http://n64devkit.square7.ch/n64man/os/osSpTaskYielded.htm

a osSpTaskYield is called but never answer as osSpTaskYielded. Graphics does not restart and the necessary parralelism of the various graphics and audio thread does not work as it should, leading to gfx errors in case of hack.

I am not sure how how the graphics plugin should behave in this case. 

Anybody has an idea?

**Comment by gonetz** [comment]
Looks like RSP plugin can yield,  but I don't see how to do it with graphics plugin.

**Comment by LegendOfDragoon** [comment]
>Looks like RSP plugin can yield, but I don't see how to do it with graphics plugin.

You'd need to modify the RSP plugin to support yielding in HLE.

**Comment by olivieryuyu** [comment]
How does it work in LLE then with PJ64  2.3?

**Comment by LegendOfDragoon** [comment]
Been a while since I did any coding / debugging for the RSP, but one issue is this https://github.com/project64/project64/blob/master/Source/RSP/Cpu.c#L216

It always does  `| 0x203` when using HLE gfx. While in LLE, if it yields, then it does not do `| 0x203`. So the RSP needs to be modified in order to support yielding in HLE. I haven't really done HLE gfx before, so there may be more work you have to do.

**Comment by olivieryuyu** [comment]
hmm it is true that LLE in GlideN64 is processing a RDP_ProcessRDPList. May be the DP status helps the matter.

in GLN64 0.3 Orkin did a small support in G_RDPFULLSYNC of the interrupt but not sure it would help.

void RDP_FullSync()
{
	*REG.MI_INTR |= MI_INTR_DP;
	CheckInterrupts();

#ifdef DEBUG
	DebugMsg( DEBUG_HANDLED, "0x%08X: 0x%08X 0x%08X RDP_FullSync\r\n", RSP.PC[RSP.PCi], RSP.cmd0, RSP.cmd1 );
	DebugMsg( DEBUG_HANDLED, "\tGenerated DP Interrupt\r\n" );
#endif
	RSP.PC[RSP.PCi] += 8;

I am getting confused with this stuff

**Comment by weinerschnitzel** [comment]
If I'm reading this right, this is an RSP issue and can be solved outside of the gfx plugin. 

1. osSpTaskYield is called, OSTask struct is saved.
2. osSpTaskYielded is called, if GFX task completes (RDP_FullSync?), return OS_TASK_YIELDED or 0x203
If GFX task NOT completed, return 0
3. osSpTaskStart is called, if we yielded (new task), reinit the OSTask struct. 
If GFX task NOT completed (osSpTaskYielded returned 0),  use saved OSTask struct.

**Comment by olivieryuyu** [comment]
yes it should RDP_FullSync as i can read.

the rdp command is described as:

Sync Full will stall the pipeline until the last command that reads or writes from the 
frame buffer finishes. This is useful for ensuring an entire scene is fully drawn before 
swapping frame buffers or reusing a frame buffer as a texture image. Before pointing the 
VI at the frame buffer in memory, ensure that this command is called so that all geometry 
appears on the screen without flicker. Due to the fact that often a program must know 
when the RDP has finished processing, Sync Full will cause a DP interrupt when the 
RDP completes processing and encounters the Sync Full. This is the only command that 
causes a DP interrupt and thus can be useful for programs that must wait for rasterization 
to complete before continuing. See the MIPS Interface (MI) registers for more details.

I guess it makes sense for Gauntlets, see the screen flickering behavior of the game without it.

**Comment by AmbientMalice** [comment]
If this behavior isn't being implemented, surely it must be having SOME effect on other games? Or is it just a really obscure feature that most devs didn't use?

**Comment by weinerschnitzel** [comment]
It sounds like devs used it to quickly switch to an audio thread. If it were to effect other games, I would look at musyx games first, and look for flickering or trap yield calls.

**Comment by olivieryuyu** [comment]
well Gauntlet Legends uses Musyx indeed. It can be an explanation.

**Comment by AmbientMalice** [comment]
>As well games like Indiana Jones/Starwars etc have some issues.

Ah, yea, that suddenly occured to me, too. Lots of "odd" behavior that very well could be audio-related.

**Comment by olivieryuyu** [comment]
yes the games in question are either using Musyx or Boss Games audio ucode (I think not HLEzied by the way). It is exactly the issue that we have a yield issue between audio and gfx.

Now question is how to implement this in our HLE environment...

**Comment by LegendOfDragoon** [comment]
Gauntlet is the only Musyx game I know of that infinite loops if yielding isn't implemented.

Some Boss games don't have an audio microcode and just use cpu instead (Twisted Edge, Top Gear Rally, WDC, and Stunt Racer).

Anyway, yielding isn't something that's easy to implement, especially since most N64 emulators are single threaded and nowhere near cycle accurate. I'd only worry about it for the games that absolutely need it (Gauntlet, WDC, Stunt Racer, etc).

**Comment by olivieryuyu** [comment]
of course, now question is how to make it happen :)

Even a hack may fit (in the gfx or the rsp or both), we speak only of few games at the end of the day.

**Comment by LegendOfDragoon** [comment]
Gauntlet uses F3DEX, so it will need a game-specific hack.

First thing that needs to be done is modify RSP to support yielding for HLE gfx plugins. If more needs to be done, then you can just debug the game with LLE to figure out the rest :smile: . Doing hacks is simple.

**Comment by olivieryuyu** [comment]
are you able to help LegendOfDragoon to modify RSP to support yielding for HLE gfx plugins? 

Seems you have a good idea on how to do it :)

**Comment by LegendOfDragoon** [comment]
You just need to find a way for the RSP to know that the HLE plugin yielded. Off the top of my head, idr which registers the gfx plugin has access to. Once you figure out what variable to use for yield detection, just add an if statement at the place i pointed to in RSP source, so that it only does `| 0x203` if it wasn't yielded.

**Comment by olivieryuyu** [comment]
A bit of documentation, even if someone would like to tackle the matter:

0x0080 signal 0 (YIELD request)
0x0100 signal 1 (YIELDED)
0x0200 signal 2 (RSP task done)

Requesting a Yield
An application requests an RSP task to yield by calling osSpTaskYield(). This function sets the Coprocessor 0 Status Register bit SP_SET_YIELD.

Checking for Yield
The microcode checks periodically for a yield request. It would be inefficient to check too often, but it would also be dangerous to not check often enough, possibly detecting the yield too late.

For the released graphics microcode, we check for the yield after processing every display list command. The test is relatively cheap, only a few cycles, and this guarantees that we will test every several hundred clock cycles at the most.

Yielding
The microcode’s responsibility during yield is, by design, minimal. The microcode saves a handful of important registers to DMEM, then DMA’s the necessary portion of DMEM to the yield buffer (originally
supplied to the task as part of the task header). The microcode also sets the SP_YIELDED bit in the Coprocessor 0 Status Register.

Saving a Yielded Process
After requesting a yield, the host CPU must wait for the RSP task to finish and verify that it actually yielded. It might also modify internal state, so that the yielded task can be restarted.

Restarting a Yield Process
Restarting a previously yielded task is conceptually simple; the previously-saved DMEM data (from the yield buffer) is used as the ucode_data field in the task header, and the OS_TASK_YIELDED bit in the
task header is set.

The microcode will detect the OS_TASK_YIELDED bit in the task header flags and perform the proper initialization, before resuming execution. This initialization should include restoring registers (from the saved DMEM) and possibly overlaying code segments.

Not sure that even LLE plugin does all of this by the way.

**Comment by olivieryuyu** [comment]
is the DP status and the SP status linked?

DP status can be accessed i think as per the zilmar specs. It may be a possibility.

**Comment by AmbientMalice** [comment]
I'm just a dummy who doesn't know much, but what if DK64's timing issues are actually an RSP yielding issue?

**Comment by oddMLan** [comment]
Hm. That doesn't sound TOO crazy tbh. Disabling the audio thread in Donkey Kong 64 fixes the graphics.
https://github.com/project64/project64/issues/791

**Comment by LegendOfDragoon** [comment]
>I'm just a dummy who doesn't know much, but what if DK64's timing/tearing issues are actually an RSP yielding issue?

I doubt that's the main issue. Tweaking counter factor in 1964 is enough to make DK not miss the swing. Not that I think 1964's timing is more accurate. I like how 1964 increments the cycle count based on the instruction rather than a universal number.

**Comment by Frank-74** [comment]
Accurate rounding needs to be done first.  We ALL need to have accurate rounding, so that everything works as expected.  Without accurate rounding, everything is a guess, or hack to work around an un-accurate base.

**Comment by Frank-74** [comment]
I just got an error with Gauntlet, switching from z64gl to GLideN64.
SP DMA WRITE
could not fit copy in memory segment.

And another one switching from GLideN64 to PJGlide.
SP DMA READ
SP_DRAM_ADDR_REG not in RDRam space.

**Comment by gonetz** [comment]
Do you run GLideN64 with fb emulation enabled? It should be disabled to avoid improper RDRAM modifications.

**Comment by olivieryuyu** [comment]
as per this discussion https://github.com/cxd4/rsp/issues/8 , it seems indeed related to yielding. Each time there is yield request, a fullsync should occur imo (does it has an effetc?). Not sure this is the case in the current hack of glide64. I guess this yield story is to swap between buffers. 

Now it can be also a core issue with the emulator (or gfx plugin, rsp and core emulator)

**Comment by tj90241** [comment]
FYI, this runs great on CEN64 (caveat: it requires the use of a PAL PIFROM to boot the cartridge).

This at least rules out the possibility of it being one of the unimplemented graphics command (i.e., tile syncs, etc.). It is probably the RSP yielding issue, or unlikely but also possible, something related to the synchronization of the CPUs.

Furthermore, it may be possible leverage CEN64 to prove that it is related to yielding: either by changing the implementation, or dumping out state that shows CEN64 is taking a different path.

**Comment by Papermanzero** [comment]
What is the current status? As far as I know, yielding seems not the root cause for the HLE issues.

**Comment by fzurita** [comment]
This is relevant here:

https://github.com/mupen64plus/mupen64plus-core/pull/257

**Comment by loganmc10** [comment]
Yeah based on my understanding, there is nothing you could do in the GFX plugin to fix this issue.

You can read a fairly good explanation of this issue here:

https://code.google.com/archive/p/mupen64plus/issues/614

I'm not sure if it would be possible for an HLE RSP plugin to handle this, but AFAIK only Project64's RSP and cxd4's RSP handle this situation currently.

**Comment by Papermanzero** [comment]
LLE would be already a start. If the devs know how the behavior is working in LLE you can possibly shift it to HLE.

**Comment by loganmc10** [comment]
> LLE would be already a start. If the devs know how the behavior is working in LLE you can possibly shift it to HLE.

Maybe, but it's possible that the HLE RSP plugins don't keep track of the information that is needed. I doubt they do.

For instance, in HatCat's RSP, he specifically disabled this behavior if the RSP is running in HLE mode:
https://github.com/cxd4/rsp/blob/master/su.c#L60-L61

I'm not sure about how Project64's RSP works.

Besides Gauntlet Legends, I think all the games that are in need of this feature require LLE anyway for the graphics.

**Comment by olivieryuyu** [comment]
@gonetz 

I begin to see what is going on but no idea how to solve the matter.

Something strange happens with the display list. When 0xDE010000 (gsSPBranchList) is used, the new displayed list is retrieved from memory but the very first command of this display list is the exactly same 0xDE010000 (with the same memory address!).

You can imagine the result: the code loops to itself!!

No idea what to do here, I wonder how it can work in LLE.

Potentially you could try to fix the matter by checking if the last instruction and the first instruction of the displaylist when using DE010000. When it is the same, the next command should be read instead.

**Comment by fzurita** [comment]
The fix I did is actually checking for this condition:

https://github.com/gonetz/GLideN64/tree/fzurita-let_gauntlet_legends_boot

I took that fix from Glide64. It's supposed to be some kind of RSP yield issue, I think the code is supposed to loop until told to break by some external source....

**Comment by loganmc10** [comment]
In LLE it works basically like this:

Normally, when the RSP completes a task, it sets the SP_STATUS_TASKDONE, SP_STATUS_HALT, and SP_STATUS_BROKE bits, then it returns control to the CPU.

If the RSP detects that it's stuck in a loop, it exits with all of those bits (taskdone, halt, broke), set to 0. This a "signal" to the CPU/emulator that the task is not done, and that the CPU should do it's thing, and then resume the same RSP task.

Does this make sense @olivieryuyu ? The problem is that the graphics plugin (with our current plugin spec), doesn't have the ability to set those bits, it is done by the RSP plugin. But in HLE this infinite loop happens in the graphics plugin. So you need some way for the graphics plugin to notify the RSP plugin that the task is not done (we are stuck in a loop), and then the RSP plugin can exit with taskdone/halt/broke set to 0.

This is my understanding. There may be some other way to accomplish this task in another method, but this is how it works in LLE currently.

**Comment by LegendOfDragoon** [comment]
>I took that fix from Glide64.

I was just about to say Glide64 already does something like that :smile: . Do you get flickering in HLE?

**Comment by loganmc10** [comment]
I tested @fzurita's fix and I did get flickering. Not in the menu, but once you start playing the game, the flickering starts.

**Comment by LegendOfDragoon** [comment]
>I tested @fzurita's fix and I did get flickering. Not in the menu, but once you start playing the game, the flickering starts.

Sounds like my experience with testing Glide64 the other day. Does it flicker in LLE?

**Comment by loganmc10** [comment]
No it doesn't

**Comment by Frank-74** [comment]
How does Rice deal with it?  It's using hacks I know, but it doesn't flicker in game with HLE.  But the status bar graphics and menu's in game are messed up.

**Comment by olivieryuyu** [comment]
@loganmc10 

Thanks for the explanation. I would expect this for WDC/Stunt Racer as a microcode feature. But  Gauntlet legends seems different.

I am totally sure of what is going, I have few theories. 

1. FIFO has to be depleted before the sub-display list can be executed.
2. Cache involvement (see http://n64devkit.square7.ch/pro-man/pro11/11-01.htm)
3. Yielding but I don't see how.

Now checking what is going on, the code seems to loop with itself but after a significant number of loops, the display list retrieved from memory does not have anymore the DE010000 at its start but a nice 0x00000000 0x00000000. Then the next command is executed and the game runs.

What could be tried is the following:

In the case there is a DE010000 as command, check if the next command is the same, if yes then check again the memory where the displaylist is located (second word of the command) until you reach a 0x00000000 and the execute next command.

That would match exactly the behavior of the code and then **may be** fix the whole flickering issue. 

@loganmc10 and @gonetz 

Would you like to give it a  try?

**Comment by fzurita** [comment]
I tried copying Rice's hack instead of Glide64's, but the game still flickers.

**Comment by Frank-74** [comment]
Is there anything related to RSP yielding in Rice?  Just seen an uint32 yield_data_ptr in RSP_Parser.h (struct OSTask_t).  Haven't looked much further though.

**Comment by fzurita** [comment]
yield_data_ptr  seems to be defined but not used in Rice.

**Comment by fzurita** [comment]
This is interesting, if I enabled the option to "Render 2d elements at N64 resolutions", most of the graphics bugs in that game go away. Also, the flickering actually ends up going away about 30 seconds into the game.

**Comment by olivieryuyu** [comment]
@fzurita @gonetz 

This option certainly delay slightly the graphics processing, leading to match the timing between the CPU and the RSP.

My understanding of the issue is that the processing of the display list get stalled with those loops. The CPU/RDP/Memory may be affected by other stream works and at one point in time, the display list get released to go on.

Those loops are simple another way to have the CPU and RSP synchronized. The plugin should check when a DE010000 occurs that the next instruction is not the same instruction DE010000. If yes the display list will get into a loop. Skipping the instruction will work to get the game running but then the CPU and the RSP will be not synchronized and the flickering will start.

As you know, the display command DE010000 retrieves information from memory in RDRAM at the location provided in the 2 half of the command.

0xDE010000
0xE0213138

0xE0213138 is the location in DRAM where the display list is downloaded. 

By looping into the same 0xDE010000 0xE0213138, you keep downloading the same displaylist over and over, where the first instruction in this new display list is 0xDE010000 0xE0213138.

However at one point in time, it will not be 0xDE010000 0xE0213138 but 0x00000000 0x00000000.

Then you need to skip this 0x00000000 0x00000000 and start running the next command in the display list.

I checked the matter more on the CPU side and at one point in time, the cache is impacted. Most likely it is what is described in the last paragraph here http://n64devkit.square7.ch/pro-man/pro11/11-01.htm.

So the plugin must checked at each loop whether the first instruction is the same DE010000 in RDRAM. When it gets to zero, the display list can be processed.

**Comment by fzurita** [comment]
I added some loving to the code. When the issue happens, the first instruction seems to always be 0xDE010000, no matter how long I let it run. I'm not sure how it could ever be updated to 0x00000000 if the CPU is not allowed to run?

**Comment by gonetz** [comment]
Good point. May be let it run:
```
	*REG.MI_INTR |= MI_INTR_DP;
	CheckInterrupts();

```
?

**Comment by olivieryuyu** [comment]
I do my analysis in LLE interpreter mode. In this mode, the ucode loops to itself a lot of time but at one point in time, something happens for sure and the 0x00000000 0x00000000 appears.

I dunno how it happens though. Interesting but tough case!

**Comment by olivieryuyu** [comment]
I digged further the case. As suspected, the game uses a LOT the CPU cache. No other game does this I think. 

In PJ64 CPU logs, from my current understanding, the case occurs when there is two addresses of sub-displayed list dram address are with cached with operation 21 at the same time.

This is what the manual describes in 11.1.3 Interaction Between the RSP and N64 CPU Memory Caching.

Zero idea how to solve this now :(

**Comment by fzurita** [comment]
@gonetz Setting ```RSP.halt = TRUE;``` and 

```
*REG.MI_INTR |= MI_INTR_DP;
	CheckInterrupts();
```

when the loop is detected, allows the game to continue. I tried looking for an address for 0x0 for w1 while this is happening, but it never happens, so the skip operation that I added is never performed, which means that the game is resuming with just a black screen since no display lists are being processed. I wonder if there is some way to determine when we should end the loop without having that 0x0 there.

**Comment by loganmc10** [comment]
If you exit the loop and hand control back to the CPU/emulator, it is going to assume that the current task is fully complete. So when the CPU/emulator gives control back to the graphics plugin, it is going to start on the next task, so you never fully completed the previous task. This is my understanding. This is why I think the things I mentioned here: https://github.com/gonetz/GLideN64/issues/111#issuecomment-312769239 matter.

Gauntlet Legends didn't work in LLE in mupen64plus until that stuff was implemented.

**Comment by fzurita** [comment]
Yeah, I get your point. It seems that the API specification needs to be extended to allow the graphics plugin to set those bits.

**Comment by fzurita** [comment]
It doesn't look that hard.

 In here:

https://github.com/mupen64plus/mupen64plus-core/blob/master/src/api/m64p_plugin.h#L69

Add a second version of ```ProcessRdpList``` with a return value or additional parameters to allow the graphics plugin to set the bits.

In here:

https://github.com/mupen64plus/mupen64plus-rsp-hle/blob/13770ff89980a91ba02d234b6426c2ea033dbfa1/src/plugin.c#L96

Check for the second version of the function, if it exists call that one, if it doesn't, call the other one.

And finally in here:

https://github.com/mupen64plus-ae/mupen64plus-ae/blob/0eecfe60905f44c603628120887dc9d942ce6a52/app/src/main/jni/mupen64plus-rsp-hle/src/hle.c#L162

Set hle->sp_status to the return value of the function instead of always setting all 3 bits to true.

**Comment by loganmc10** [comment]
Yeah it wouldn't be terribly difficult. Like you say, instead of the "ProcessDList" command returning void, you have it return ```false``` if the task needs to be resumed, or ```true``` if the task is completed, something like that. Then the RSP plugin sets the bits accordingly.

**Comment by fzurita** [comment]
That should work with mupen64plus, no idea about how involved it would be to do it with Project 64 though.

**Comment by loganmc10** [comment]
Yeah, and I don't know if the RSP plugins do any other magic besides setting those bits to make this work. Someone like zilmar or cxd4 who wrote the RSP plugins should probably know, but I'm not 100% sure if this will work.

**Comment by fzurita** [comment]
Me neither, I don't think it's that much code to do what I said. So I'll try it tonight and if it doesn't work, who cares, it's not that much work.

**Comment by olivieryuyu** [comment]
short question: do you expect the plugin to set the bytes for the yielding?

Also my cache theory is not valid?

**Comment by fzurita** [comment]
Yeah, I would estimate the plugin to set status bits. I'm not sure I truly understand your cache theory to have much input on that. I will need to read more about it to increase my background on it.

**Comment by olivieryuyu** [comment]
well the issue is that the RSP does not change set the status bits when 0xDE010000 is met, hence the fact that I would not understand why implementing yield for this game makes it working. But you can of course give it a try. I would think that it will be necessary for WDC/Stunt Racer anyway.

and your question is correct: why it works in LLE with cache?

**Comment by fzurita** [comment]
Here is some test code I did: https://github.com/mupen64plus-ae/mupen64plus-ae/commit/74a83dde808a02c971e805015dd18ff315648e40

Unfortunately, this gives me a black screen in Gauntlet Legends. Other games. work fine. I must be doing something wrong.

**Comment by loganmc10** [comment]
I think maybe we are getting our terminology mixed up. When I say "yield" I mean "pause the RSP, let the CPU run for a while, then resume the RSP where it left off". I know the N64 has a "task yielded" bit/instruction, but that is not what I'm talking about.

**Comment by fzurita** [comment]
Ok, I found a couple of problems in my last implementation and I updated my branch. It's actually working now without skipping the instruction! Although, it's really really slow, it takes forever for the instruction to turn into a non-looping instruction.

mupen64plus-ae/mupen64plus-ae@74a83dd

**Comment by olivieryuyu** [comment]
Great!

We are nearly there!!

**Comment by fzurita** [comment]
Ok, I think I got it working without hacks thanks to all the hints from @olivieryuyu and @loganmc10 . Unfortunately, the flickering is still there in-game. So either I'm not doing it 100% right or it's a custom LLE ucode.

**Comment by olivieryuyu** [comment]
It is not a custom ucode as far as I can see.

Now i got an idea: the RDP commands stored in the fifo are only actually send to the RDP after the the sub display list start running (the one 0xDE010000). In HLE the commands are immediatly executed but in LLE only when received by the RDP. This matter explains the difference between HLE and LLE and explains that the display has a potentially a timing issue.

@gonetz 

Is there a way to emulate the fifo?

**Comment by fzurita** [comment]
@olivieryuyu Do you know what the CPU is doing while the RDP is in a loop?

**Comment by olivieryuyu** [comment]
Unfortunately no.

**Comment by fzurita** [comment]
I have the feeling that we don't want to swap buffers while the RDP is in a loop.

**Comment by olivieryuyu** [comment]
one interesting thing to know is that 0xDE010000 is the gSPBranchList and the issue occurs only when the stack is empty. But unsure why it would cause an issue.

**Comment by theboy181** [comment]
Can you up a version of GLideN64 that shows the HLE flicker in this game?

**Comment by theboy181** [comment]
Is there any chance anyone here has a zilmer spec version of this available for testing?

**Comment by fzurita** [comment]
I can make one tonight. I've just been busy.

**Comment by theboy181** [comment]
Thanks...

**Comment by fzurita** [comment]
@theboy181 Here you go, against the latest master:
https://drive.google.com/file/d/0B57Ioy26LWegN0g0NGx5SWdSVUU/view?usp=sharing

**Comment by theboy181** [comment]
Its seems to be related to up and down c presses.. The screen draws the menu in the wrong cords.

**Comment by Frank-74** [comment]
The flickering stops after about 10 seconds here.  But I cannot enter the first level.  It crashes every time I try.

**Comment by olivieryuyu** [comment]
I can get into the game. It stops but as soon as you are hit, it start again. The framebuffer is updated but keeps flickering. Weird stuff!

**Comment by Frank-74** [comment]
Tried to get into Mountain Valley about two dozen times.  PJ64 always crashes.  Toggled just about every option in GLideN64 and PJ64 to get into the first level.  Tried CPU interpreter and RSP interpreter as well.
My wizard spins into the portal, screen goes black and PJ64 crashes.

**Comment by theboy181** [comment]
@Frank-74 I would try an earlier build of PJ. The OC version crashed with error message, and a random one I had in my folder 2,4,? loaded, but was a black screen with the audio.  Maybe a CF issue?

**Comment by theboy181** [comment]
@olivieryuyu what emulator are you using?

**Comment by olivieryuyu** [comment]
Mupen 0.5, PJ 2.3

**Comment by fzurita** [comment]
Keep in mind that the only change I did for the zilmar version was this since I'm not as familiar with the API:

https://github.com/gonetz/GLideN64/commit/e49be85ac4daeaa368c934e5dc2e9793fac1f320

**Comment by Frank-74** [comment]
@theboy181 Tried an earlier PJ64 version from April.  It's the same.
Tried again with a version from February, it crashes before spinning into the portal.
I'm not getting any error messages, it's an application crash.

**Comment by theboy181** [comment]
@olivieryuyu does PJ2.3 allow you to start the first level? @Frank-74 are these versions you mentioned 2.3?

**Comment by Frank-74** [comment]
@theboy181 No, but since then, I went back to 2.3.1-234 build.  From 25/11/2016.
Been progressing through earlier versions.  Still the same crash.

**Comment by theboy181** [comment]
@fzurita could the issue be related to how Gauntlet HLE handles interlacing? I know that AL had to address the issue with his LLE plugin. Different video timings caused issues with the health bar, but it is now fixed since he examined the code. Maybe @LegendOfDragoon can share the PR that fixed it?

**Comment by fzurita** [comment]
@theboy181 Hmmm, that's past my area of expertise, but I could see how if the game was producing an interlaced signal with only odd frames containing image data, I could see the flickering happening. Maybe the answer is to skip swapping buffers for those frames if they can be identified within the code some how. This may not be hard since it seems that now draw commands are performed for those bad frames (the image is all black).

**Comment by olivieryuyu** [comment]
Do you mean the VI_V_CURRENT_LINE_REG? Not sure indeed that GlideN64 handles this reg.

**Comment by LegendOfDragoon** [comment]
I think this was the commit
https://sourceforge.net/p/angrylions-stuff/code/99/tree//trunk/mylittle-nocomment/n64video.cpp?diff=55bb05dc27184618d775bec1:98

**Comment by fzurita** [comment]
Interesting, enabling multisampling causes black textures in this game.

**Comment by fzurita** [comment]
Ok, finally! I got the game working without flicker. I can also enter the first level without issues in HLE mode. This only works with mupen64plus though, here are the code changes so far:

https://github.com/mupen64plus-ae/mupen64plus-ae/tree/gauntlet_legends

The only problem I see so far is that the game seems to be running a little slower that it should even at a decent frame rate.

**Comment by olivieryuyu** [comment]
Woa nice! 

However getting LLE command here is odd. Potentially it is how it should be, yet it is puzzling.

May be you avoid the issue without solving it (no offence, it is just that I am puzzled by this LLE matter)

On my side, I have just noticed some strange command DE000000 01000000 or DE010 02000.There is not rdram address. However after a routine which normally only suppose to fix the address, it gets an address. I wonder if it can be the root of the issue.

But very nice result so far, we are getting closer to something.

**Comment by fzurita** [comment]
The weird LLE commands went away after my last commit. I must had been doing something wrong previosly.

**Comment by olivieryuyu** [comment]
ok so which commits are actually really necessary? 

https://github.com/mupen64plus-ae/mupen64plus-ae/commit/11e1d63e35e582f17e3d0c08a3633ec360ba20af

and https://github.com/mupen64plus-ae/mupen64plus-ae/commit/167f9191cd7b5d5e69e935e2c188a75bb13466fa

only?

**Comment by fzurita** [comment]
These two:
https://github.com/mupen64plus-ae/mupen64plus-ae/commit/11e1d63e35e582f17e3d0c08a3633ec360ba20af
https://github.com/mupen64plus-ae/mupen64plus-ae/commit/787c549269d6d1faa0003af137155f1a112c23e5

**Comment by olivieryuyu** [comment]
You can be right, just changing values is more safe!

I am happy to see that no LLE is needed, would have been weird!

**Comment by loganmc10** [comment]
Just tested this on my laptop and I can confirm it works! But ya, it's really slow, and my CPU isn't maxed out or anything.

What exactly should I test? whether addr0 or addr1 is different from their previous values? If either of them have changed then ```return FALSE```?

**Comment by loganmc10** [comment]
Also the HUD (bottom left) looks messed up, but that might be unrelated to what we are dealing with here (maybe it's just an unrelated HLE bug?)

![gauntlet_legends-000](https://user-images.githubusercontent.com/848146/28177508-3ecaf5b6-67b8-11e7-8110-690059ef6d65.png)

**Comment by fzurita** [comment]
Yeah, I was going to try just checking for a change in addr0 or addr1. As long as addr0 and addrs1 remain the same (we are looping) then keep on returning FALSE.

**Comment by Frank-74** [comment]
@loganmc10 Latest angrylion displays health bar correctly.
![glangrilion](https://user-images.githubusercontent.com/11056843/28177890-42765b1a-67f4-11e7-9666-a5473bd3ef5b.JPG)

**Comment by loganmc10** [comment]
I just tried this:
```
if(addr0 == old_addr0 || addr1 == old_addr1)
{
    //LOG(LOG_ERROR, "%08x (w0:%08x, w1:%08x): ", RSP.loopAddress, w0, w1);

    return FALSE;
} else {
    old_addr0 = addr0;
    old_addr1 = addr1;
    RSP.inLoop = false;
}
```

It works, but it's just as slow.

@Frank-74 yeah I just realized that the health bar/HUD looks messed up in GLideN64 LLE as well, so it's unrelated to this issue of "yielding"

**Comment by LegendOfDragoon** [comment]
That HUD issue seems related to VI emulation. As @theboy181 pointed out, it was also an issue in older version of Angrylion's.

**Comment by Frank-74** [comment]
I built the latest angrylion from here: https://sourceforge.net/p/angrylions-stuff/code/HEAD/tree/trunk/mylittle-nocomment/

I changed the resolution to 640x480 in my build.  1024x768 was too big windowed on my laptop.

**Comment by fzurita** [comment]
I'm not sure what is happening, but the game is running at around 4 fps even when the core is running at 60VI/s. But at least the flicker is gone... lol

**Comment by fzurita** [comment]
I got it running full speed with my change and no flicker. Set count per op to 1 and count per scanline to 2100.

**Comment by theboy181** [comment]
Has this been tested on pj64 yet?

**Comment by fzurita** [comment]
It requires making changes to the API. I'm not as familiar with the zilmar API to make the same changes there.

**Comment by Gillou68310** [comment]
@fzurita the main problem with your implementation is that the task is restarting from the beginning instead of being resumed from the last executed display list command.

Here's how I would have done this:
https://github.com/Gillou68310/mupen64plus-core/tree/gauntlet
https://github.com/Gillou68310/mupen64plus-rsp-hle/tree/gauntlet
https://github.com/Gillou68310/GLideN64/tree/gauntlet

Could you check if my implementation is correct?

**Comment by fzurita** [comment]
Sure, I'll check some time tonight.

**Comment by Gillou68310** [comment]
I gave access to the status register within the video plugin. At least we now have access to the yield flag in case we ever need it.

**Comment by Gillou68310** [comment]
I forgot to bump the API version

**Comment by loganmc10** [comment]
@Gillou68310 I tried yours out, it seems to behave the same as @fzurita's version. It's quite slow, but like @fzurita mentioned, if I set CountPerOp=1, it speeds up. It actually seems a little too fast to me, but I don't have much experience in this game so I'm not sure what it should sound like.

The CPU usage is also really high with CountPerOp=1

**Comment by fzurita** [comment]
Try playing around with CountPerScanline as well.

**Comment by loganmc10** [comment]
Yeah it seems better with CountPerScanline=2200, it makes the audio less crackly. It definitely works though, no flickering or anything. I just wonder why the CPU usage is so high

**Comment by fzurita** [comment]
Not sure, CPU usage is much lower when using the normal hack to just skip the instruction.

**Comment by fzurita** [comment]
@loganmc10 do you get flickering if just using the hack but changing CountPerOp and CountPerScanline?

**Comment by loganmc10** [comment]
Yeah it still flickers with CountPerOp at 1 or 2, and CountPerScanline at 1500 or 2200

**Comment by fzurita** [comment]
So this approach does seem legit, we just need to figure out why performance doesn't seem right.

**Comment by standard-two-simplex** [comment]
I would guess that with the new condition, False is returned less often, so the CPU has to go through all the rest of the code much more often.

**Comment by olivieryuyu** [comment]
One thing I noticed is that during the loops, the cpu seems to work on its cache. I m no expert on CPU, may be nothing important.

**Comment by Gillou68310** [comment]
Thanks for testing I'll see what I can do about the perf issue. LLE doesn't have this issue right?

**Comment by fzurita** [comment]
It doesn't. It actually runs faster in LLE mode if you don't change the CountPerOp and CountPerScanline settings.

**Comment by fzurita** [comment]
@Gillou68310 One thing I noticed about the way you did the API change is that I don't think it will work with older video plugins that don't support the new API, while my version should still work.

**Comment by Gillou68310** [comment]
Yep the only requirement is that older video plugins should be recompiled with the new m64p_plugin.h

**Comment by fzurita** [comment]
I can probably use CXD4 in LLE to monitor the register during the loop to make sure we are behaving the same in HLE.

**Comment by Frank-74** [comment]
Looks easy enough to make equivalent change to Project64.
I copied Gillou68310's api change here: https://github.com/Frank-74/project64/commit/4b8a7e027c8f8295577b4da27b7a8b89203adbdb

Don't know if it works as I need a zilmar spec version of GLideN64 with the changes to test.

**Comment by loganmc10** [comment]
@Frank-74 you'd probably need to make some changes to the RSP plugin as well. Right now, it (likely) sets ```SP_STATUS_TASKDONE | SP_STATUS_BROKE | SP_STATUS_HALT``` after getting control back from the graphics plugin. What @Gillou68310 did is have it set those bits before the graphics plugin takes control, and now the graphics plugin can unset them if it needs.

**Comment by Frank-74** [comment]
@loganmc10 I can only find relevant things in the Android RSP.  There's no rsp_break or hle_execute in the win32 version of the rsp that I can find, or any other similar references to SP_STATUS_BROKE | SP_STATUS_HALT either.

**Comment by loganmc10** [comment]
```*RSPInfo.SP_STATUS_REG |= (0x0203 );```

0x0203 is ```SP_STATUS_TASKDONE | SP_STATUS_BROKE | SP_STATUS_HALT``` (taskdone is 0x0200, broke, is 0x0002, halt is 0x0001)

So basically to mimick what @Gillou68310 did, you want to move ```*RSPInfo.SP_STATUS_REG |= (0x0203 );``` to before ```RSPInfo.ProcessDList();```

**Comment by fzurita** [comment]
There are big differences in the SP_STATUS_REGISTER between HLE using RSP-HLE and LLE using CXD4. I suspect RSP-HLE has bugs.

Here is the pattern in LLE mode during the loops, and it's very simple:
```
0x40 = SP_STATUS_INTR_BREAK
0x243 = SP_STATUS_HALT, SP_STATUS_BROKE, SP_STATUS_INTR_BREAK, SP_STATUS_TASKDONE
```

Here it is in HLE:
```
0xc0 = SP_STATUS_DMA_BUSY, SP_STATUS_DMA_FULL
0x2c3 = SP_STATUS_HALT, SP_STATUS_BROKE, SP_STATUS_DMA_BUSY, SP_STATUS_DMA_FULL, SP_STATUS_TASKDONE
0x243 = SP_STATUS_HALT, SP_STATUS_BROKE, SP_STATUS_INTR_BREAK, SP_STATUS_TASKDONE
0x40 = SP_STATUS_INTR_BREAK
0xc0 = SP_STATUS_DMA_BUSY, SP_STATUS_DMA_FULL
```

**Comment by theboy181** [comment]
Found an interesting link to the flickering. in AL when you watch the health bar you will see that is slowly brightens till you can see BRIGHT RED on the edges.  The flickering when using GLideN64 with the original hack, it stops after the last RED update is made. 

I am certain the issue is linked to how the VI is drawn to the screen. @gonetz should look at how the VI was fixed in the link AIO provided earlier.

**Comment by theboy181** [comment]
An interesting observation with the flickering in Gauntlet, while using GLideN64. 

Using AL RDP when you watch the health bar you will see that it slowly brightens till you can see BRIGHT RED on the edges of the health bar.  The flickering when using GLideN64 with the original hack, stops after the the last update of the bars brightness is made. I believe this is a clue, and it may indicate how the screen is updating the screen.

Is the issue is linked to how the VI is drawn to the screen? @gonetz should look at how the VI was fixed in the link @LegendOfDragoon  provided earlier. I know I reported the issue to AL over a year ago, and I suspected the issue was related to improper interlacing technique.

**Comment by LegendOfDragoon** [comment]
The HUD was actually fixed in r96, for Angrylion's :smile: . Same commit that fixed Mischief Makers.

**Comment by theboy181** [comment]
![capture](https://user-images.githubusercontent.com/9299743/28242013-861c8062-6955-11e7-90fd-806b77008a46.PNG)

Looks like the bug was fixed in r96 of AL's RDP. However I used CDX4's current AL's RDP OpenGL 1.5 and too the screen shot above.

I think the r96 fixed the rocks not showing up Mischief Makers, and at that time corrected the issue in Gauntlet. I guess it wasn't an interlaced issue. My reasoning for this assumption was my AL RDP with scan lines code altered the interlacing to achieve its scan lines, and the bug didn't happen with scan lines were enabled. If anyone thinks that its linked and wants me to upload the build for that, let me know.

**Comment by fzurita** [comment]
I tried the same RSP modifications in CXD4 running in HLE video mode, but they don't work for some reason.

**Comment by olivieryuyu** [comment]
r96 was implemented in GlideN64 so hard to believe it is the root cause of the issue.

**Comment by fzurita** [comment]
I got it working with CXD4 RSP in HLE mode, but the bad performance is still there. Only thing I can think of is that CXD4 RSP is not doing something in HLE video mode that it is doing in LLE mode.

**Comment by LegendOfDragoon** [comment]
can someone link the commit in GLideN64 that fixed the rocks in Mischief Makers?

**Comment by gonetz** [comment]
https://github.com/gonetz/GLideN64/issues/719#issuecomment-181929826

**Comment by olivieryuyu** [comment]
@loganmc10 

I am checking a bit again the game.

What is the SP status just before and when the loops ends?

Would it be 0x80 by any chance?

**Comment by loganmc10** [comment]
@olivieryuyu I assume you're asking about this sort-of-working HLE implementation? I used Gillou68310's version to test.

While it's "stuck in the loop" (the RSP task returns with HALT and BROKE set to 0), the value of SP_STATUS_REG is 0x40 (before and after the RSP task runs) for a while, then it changes to 0xC0 for a while.

When it finally finished the RSP task, SP_STATUS_REG is 0xC0 before the RSP task runs, when the RSP task is done, SP_STATUS_REG is 0x2C3

I don't know if that made sense, it looks like this basically:

run RSP task (0x40 before and after the task)
run RSP task (0x40 before and after the task)
...this happens for a while
run RSP task (0xC0 before and after the task)
...this happens for a while
run RSP task (0xC0 before task, 0x2C3 after, task finished!)
run RSP task (0x40 before task, 0x243 after, task finished!)
run RSP task (0x40 before task, 0x243 after, task finished!)
run RSP task (0x40 before task, 0x243 after, task finished!)
run RSP task (0x40 before and after the task)
repeat.....

**Comment by olivieryuyu** [comment]
@loganmc10  

Thanks!!! 

I think i did not express myself correctly or I do not understand your comments

What is the SP status just before the displaylist stops looping?

What is the SP status just before the displaylist start again (meaning without the same instruction over and over).

In the microcode there is potentially one possibility to do something different than what it is doing during the loop and it would mean that if you AND the SP status register by 0x80, it would be different than zero (so 0xC0 and 0x2C3)

**Comment by olivieryuyu** [comment]
And within the loops, do you have other 0xC0?

**Comment by loganmc10** [comment]
In LLE it's quite different, usually it's 0x40 just before it stops looping as well as just before it starts again, but I do see a few instances (regularly), where it is 0xC0 just before it stops, and when it stops the status is actually 0x3C3, then it's 0x40 before it starts again.

Hopefully that made more sense

**Comment by loganmc10** [comment]
> And within the loops, do you have other 0xC0?

yes, it within the loop it stays at 0xC0 for quite a while, but remember, this is with the sort of broken HLE implementation.

In LLE, it's only 0xC0 once at the beginning before the loop exits, then when the RSP task is done, the status becomes 0x3C3, then it starts a new RSP task at 0x40.

**Comment by loganmc10** [comment]
So maybe the missing step in the HLE implementation is to exit the loop if SP_STATUS & 0x80 is true, is that what you're thinking?

**Comment by olivieryuyu** [comment]
In LLE 0xC0 just before the exit of the loop!?!!

I need to check few things but we have a new lead!! 

Thanks!!!!

**Comment by theboy181** [comment]
How about this?  https://1drv.ms/u/s!AlHyj2CfwNtQiA5b6lTWO9p4JKxr

**Comment by olivieryuyu** [comment]
seems to work...

what is this version?

**Comment by oddMLan** [comment]
@LegendOfDragoon made his own attempt at solving the RSP yielding. He added special code to GLideN64 and his RSP to be able to do this

**Comment by fzurita** [comment]
I'm not familiar with what @LegendOfDragoon did. Does it fix the high CPU utilization during HLE mode?

**Comment by theboy181** [comment]
@olivieryuyu I think he just took his own advise with the yielding

**Comment by theboy181** [comment]
@fzurita seems to run well, however there needs to be a fix for the health bar to make it playable imo.

**Comment by olivieryuyu** [comment]
Impressive really happy to see that somene figure out the issue.

@LegendOfDragoon 

Well done!

So after all the arguments around the matter, was it indeed a yielding issue, a ucode issue, or what I do believe, a combination of both?

**Comment by oddMLan** [comment]
I want to know how he solved it, too.

**Comment by LegendOfDragoon** [comment]
> was it indeed a yielding issue, a ucode issue, or what I do believe, a combination of both?

It is essentially yielding. From my debugging session, I found that the flickering could not be avoided without letting the cpu run for a while. I assume that the CPU changes some of the DList data. It differs from WDC, Stunt Racer, Pokemon Stadium 2 GB Tower, in that Gauntlet makes the branching command infinite loop for the RSP to wait for CPU if it's too far ahead, instead of using the RSP yield signal. Gauntlet's method seems less efficient, but my guess is that they didn't want to tweak the microcode for that.

All I did was use a global variable that both the RSP and GFX plugin can access, that signals whether yielding happened in HLE. When the inf loop happens, signal that it exited early and exit the DList immediately. Upon resuming the yielded task, I made the code immediately jump back to the same branch function and check if it's still an infinite loop.  I could try seeing if it works in Mupen. I don't have the source with me right now though. Does M64p compile with msvc? I might need to debug their code to see why the performance is not good.

I think the main issue really is the plugin spec. I had to use a very ugly workaround where I used an unused RDP register for signaling. It would be much better to just add a variable to the spec for signalling a yield in HLE.. So i'm kinda waiting for that to happen because I really don't like this workaround.

>Does it fix the high CPU utilization during HLE mode?

There's a minor performance hit, compared to the old bypass method that had flickering, but nothing major.

Anyway, I started looking into the HUD issue. I suspect that the game loads data somewhere just once in RDRAM that deals with the HUD, before entering a different area/stage. I'm almost certain the problem is alpha blending related. The game seems to constantly fill that area (in VI_ORIGIN) with 0x0001's, every frame. I'm still debugging and will keep digging further. I'm going to try figuring out where that area of memory is for the HUD.

**Comment by fzurita** [comment]
Ok, this sounds essentially the same as what me and then @Gillou68310 did (but using a better method). We were able to get rid of the flickering, but we get really bad CPU performance.

From comparing LLE to HLE, it seems like HLE ends up spending a lot more time in the CPU. The game can run full speed with Mupen64plus on a PC if the CountPerOp setting is set to 1 and the VIRefreshRate is set to 2200. Those parameters don't have to be changed for LLE mode at all, although the game ends up running much slower in LLE because of the video plugin running in LLE.

**Comment by LegendOfDragoon** [comment]
I just now looked at Gillou's code and it looks right. Most likely an issue with Mupen then.

I'm still puzzled that HLE would spend more time in CPU.

**Comment by loganmc10** [comment]
It's probably not quite the same thing, I'd be interested to see the source code. Did you mess with the halt/broke bits at all? If not then it probably works quite differently

**Comment by fzurita** [comment]
Maybe I'm wrong, but I don't think the Mupen64plus core behaves any different between HLE and LLE, it's all up to the RSP and RDP. I'm thinking that the RSP is doing something differently in HLE that it shouldn't be. If you use the CXD4 RSP and make a similar modification there, do you still get good CPU performance?

**Comment by theboy181** [comment]
![image](https://user-images.githubusercontent.com/9299743/29246031-110b251e-7fa1-11e7-82f6-f94f3f7176ea.png)
Picture ave is from some debugging that was done with AIO. Making a save state using AL's PLug then switching to GlideN64 results in the bar rendering and working, however there is a complete line missing from the HUD when doing so. 

What was observed is that GLideN64 doesn't read from RDRAM yet all other HW renders do. Is this function not working in GLideN64 or is it just this game?

**Comment by LegendOfDragoon** [comment]
I can't test GLideN64 until later today so I decided to briefly test with Glide64. Later today, I'll test on GLideN64 in Mupen and share my findings.

>Did you mess with the halt/broke bits at all?

Only thing I did for that was skip writing to status register and RSP interrupts, if the infinite loop occurred.

>If you use the CXD4 RSP and make a similar modification there, do you still get good CPU performance?

I tried applying it to cxd4's RSP and it worked fine there too. With Glide64 i got roughly an 8% performance hit from the yielding implementation, in PJ64.

**Comment by fzurita** [comment]
If you are able to, I would really like to see your implementation.

**Comment by LegendOfDragoon** [comment]
I tried my implementation in mupen and it doesn't work. Either I messed something up in the HLE RSP, or the emulator core works differently.

Anyway I'll attach source for the changes I made in GLideN64. It certainly works in PJ64.
[gSP.zip](https://github.com/gonetz/GLideN64/files/1223229/gSP.zip)

for RSP i just did
`void hle_execute(struct hle_t* hle)
{
    if (is_task(hle)) {
        if (!try_fast_task_dispatching(hle))
            normal_task_dispatching(hle);
			
        if (hle->dpc_pipebusy == 0)
            rsp_break(hle, SP_STATUS_TASKDONE);
        else
            hle->dpc_pipebusy = 0;
    } else {
        non_task_dispatching(hle);
        rsp_break(hle, 0);
    }
}`

**Comment by loganmc10** [comment]
@LegendOfDragoon hle->dpc_pipebusy is a pointer, so you just need ```*hle->dpc_pipebusy```. Runs at full speed! great work. The audio is a little staticy in mupen64plus, it sounds a lot better with CountPerOp=1,  but still not perfect, obviously nothing to do with this though.

**Comment by fzurita** [comment]
Nice!

**Comment by AmbientMalice** [comment]
Wonderful, wonderful work guys.

**Comment by fzurita** [comment]
We just need to coordinate pull requests with all the separate projects, won't be fun.

**Comment by LegendOfDragoon** [comment]
Good to hear :smile: !

>@LegendOfDragoon hle->dpc_pipebusy is a pointer, so you just need *hle->dpc_pipebusy.

Wow haha. Normally I'd just run the emulator through a debugger to catch these kind of mistakes but I don't have a debugger installed on my desktop. I normally do most of my work on laptop. I guess I really should install MSVC 2017 on my desktop then lol.

**Comment by KingOfHyrule** [comment]
Hi! I've been watching this issue for a while, as a huge fan of this game, and noticed that you guys seem to think the colored bar in the HUD shows the player health. But this is actually wrong, as health in this game is shown as a number, right here (red circle):
![health](https://user-images.githubusercontent.com/31485246/29886160-8c7d654e-8d8f-11e7-90a4-47b3f36be358.jpg)
You can see that there's the lower tip of a heart icon cut off, and also a number right beside it. That's the health indication.

The bar you guys are looking at is, in fact, used for turbo (special/mass attacks), and updates regularly, every second or so, from green to yellow and then red. Once full, it stops and stays there, until the player uses said turbo (special/mass attack), which completely empties it and it starts filling up again, bit by bit. Each colored bar means a different power of the turbo (special/mass attack), in the following order:

1) no bar filled: no special attack possible;
2) green bar filled up: weak power attack;
3) green and yellow filled up: medium power attack;
4) green, yellow and red filled up: most powerful attack.

Now, I'm not a coder at all, and can barely keep up with the details discussed here, but it seems you guys don't understand why the game keeps sending "signals" which changes the state of the HUD all the time. Well, the only thing that isn't static in the HUD is this bar, since it keeps filling up until either the player uses the turbo/special/mass attack (which brings the bar back to "zero", and starts filling from there), or when it gets full, where it stays until the special/mass attack is used.

I believe this turbo/special/mass attack can be used by pressing Z + A (n64 buttons), but it has been nearly 15 years since I've played it, so bear with me  :)

I hope this information helps a bit on understanding what the game is doing.

**Comment by olivieryuyu** [comment]
I spent a bit more time on the game.

I begin to wonder if the game is just not badly written by the devs of this game.

Usually you finish a displaylist stored in memory by a command telling that the displaylist is over or a command that call a new displaylist, etc..

In this game the displaylist sometimes does not do so, goes beyong its maximum length in DMEM and therefore the displaylist memory stack is called to the rest of commands of the displaylist.

But even worst, the displaylist does very often this:

Texture rectangle:

E4XXXXXX
XXXXXXXX
E1XXXXXX
XXXXXXXX

at the end of the displaylist!!!

Consequence: the rest of the displaylist has to be retrieved from memory before continuing the rest of the command, meaning:

F1XXXXXX
XXXXXXXX

I clearly wonder if the N64 expects this kind of matter and if this is not the cause of the flickering!

**Comment by olivieryuyu** [comment]
The flickering in game is due what i explained above. I think we can fix this without playing with yielding! Moreover it may fix the missing gfx!

**Comment by olivieryuyu** [comment]
@gonetz 

any chance to have a small look to this please? :)

**Comment by oddMLan** [comment]
Would you like to discuss this on our Emulation 64 Discord? If you don't mind. almost everyone involved in the N64 scene is there, you would be a great addition to it :)

https://discord.gg/MF8636

**Comment by gonetz** [comment]
@olivieryuyu I currently busy with maintenance tasks. I'll definitely look at this,  but later.

**Comment by loganmc10** [comment]
> There were talks about the yielding issue at channel #zilmar, he thinks it might be not that easy "the problem is that the RSP sets registers after a dlist, that has to stop if you want yielding controled by gfx plugin"

That's not really a huge problem. Right now, after the dlist is processed, the RSP plugin sets ```BROKE | HALT | TASKDONE```. You can just change it to have the RSP plugin set ```BROKE | HALT | TASKDONE``` before the dlist is processed, then pass SP_STATUS_REG to the GFX plugin, and allow the GFX plugin to modify those bits. It would still be compatible with any GFX plugin, and will allow the GFX plugin to not set those bits if it shouldn't. That is the solution that @Gillou68310 used in his earlier attempt at this

**Comment by gonetz** [comment]
Test build from gauntlet_fixes branch with recent build of mupen64plus with updated API from loganmc10:
https://drive.google.com/file/d/0B0YqMPjGo3B2ZXlwUl9lV2dLdUE/view?usp=sharing

I tested it a bit and met no problems with this game.

**Comment by olivieryuyu** [comment]
there is still the powers graphics that are incorrect.

![untitled](https://user-images.githubusercontent.com/8987490/31535828-d4609b4e-affc-11e7-937a-8f9efa923853.jpg)

**Comment by Papermanzero** [comment]
Are the optimisations already merged (in Mupen64Plus and GlideN64)?

**Comment by loganmc10** [comment]
@Papermanzero The code is in place in mupen64plus to support this, but the RSP plugins still currently force LLE for Gauntlet Legends, you'd need to compile the RSP plugin yourself to allow it to run in HLE.

GLideN64 hasn't merged these changes into master yet, once/if that happens, then the mupen64plus RSP plugins would be modified to allow the game to run in HLE by default

**Comment by gonetz** [comment]
Merged to master: 168dccde2a3

**Comment by Papermanzero** [comment]
I updated mupen64plus and gliden64 to the newest master. For th RSP, i used loganmcˋs adapted plugin which gonetz posted. The game is running perfectly. The issue is the turbo bar. But everything else is working flawlessly. This is a major progress.

**Comment by gonetz** [comment]
>I would suggest to close the issue and open a new issue concerning the turbo bar. 

Agreed.
